tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -0,0 +1,955 @@
+  } else if (config_bmm1.alpha_real() != scale_value) {
+    scale_value = config_bmm1.alpha_real();
+  }
+
+  fmha_config.set_fmha_scale(scale_value);
+  fmha_config.set_dropout_rate(dropout_rate);
+  // Set to an arbitrary seed for now, seed is not exposed to XLA in HLO
+  // graph.
+  // TODO Find a way to compute original seed from dropout keys.
+  fmha_config.set_seed(42);
+
+  *fmha_config.mutable_intermediate_tensor_shape() = bmm_1->shape().ToProto();
+  {
+    auto* algorithm = fmha_config.mutable_algorithm();
+    algorithm->set_algo_id(0);  // engine id
+    algorithm->set_math_type(se::dnn::AlgorithmProto::TENSOR_OP_MATH);
+    std::vector<int64_t> knob_ids = /* {0, 1} */ {17, 24};
+    std::vector<int64_t> knob_vals = {1, 0};
+    for (int i = 0; i < knob_ids.size(); ++i) {
+      (*algorithm->mutable_tuning_knobs())[knob_ids[i]] = knob_vals[i];
+    }
+    algorithm->set_is_cudnn_frontend(true);
+    algorithm->mutable_workspace_size()->set_value(0);
+  }
+  const Shape& output_shape = bmm_2->shape();
+  Shape call_shape =
+      ShapeUtil::MakeTupleShape({output_shape, ShapeUtil::MakeShape(U8, {0})});
+
+  std::vector<HloInstruction*> operands = {lhs_bmm1, rhs_bmm1, rhs_bmm2};
+  if (mask != nullptr) {
+    HloInstruction* converted_mask = comp->AddInstruction(
+        HloInstruction::CreateConvert(bmm_1->shape(), mask));
+    operands.push_back(converted_mask);
+  }
+  if (bias != nullptr) {
+    HloInstruction* original_bias;
+    // There will be cases where the bias is up-casted to wider float type,
+    // we need to take the original bias node and broadcast it without
+    // converting.
+    if (Match(bias, m::Broadcast(
+                        m::Convert(
+                            m::Op(&original_bias)
+                                .WithPredicate([](const HloInstruction* instr) {
+                                  return instr->shape().element_type() == F16 ||
+                                         instr->shape().element_type() == BF16;
+                                }))
+                            .WithPredicate([](const HloInstruction* instr) {
+                              return instr->shape().element_type() == F32 ||
+                                     instr->shape().element_type() == F64;
+                            })))) {
+      std::vector<int64_t> broadcast_dims;
+      switch (original_bias->shape().rank()) {
+        case 1:
+          broadcast_dims = {3};
+          break;
+        case 2:
+          broadcast_dims = {0, 3};
+          break;
+        case 3:
+          broadcast_dims = {0, 2, 3};
+          break;
+        default:
+          broadcast_dims = {};
+          break;
+      }
+      Shape bcast_shape = bmm_1->shape();
+      bias = comp->AddInstruction(HloInstruction::CreateBroadcast(
+          bcast_shape, original_bias, broadcast_dims));
+    }
+    operands.push_back(bias);
+  }
+  HloInstruction* fmha_call =
+      comp->AddInstruction(HloInstruction::CreateCustomCall(
```

Comment:
Does this propagate metadata?

Reply:
Added metadata. It seems there is no good way to combine metadata right now in XLA. Hence, I added metadata of one of the BMMs.

Reply:
Yes, we usually take the metadata from the "root" instruction, and then propagate that.